File: app/src/api/invoices.ts
import { invoke } from "@tauri-apps/api/core";

// ── Types ────────────────────────────────────────

export type InvoiceStatus = "draft" | "sent" | "paid" | "cancelled";

export interface InvoiceLine {
  id: number;
  invoice_id: number;
  catalogue_id: number | null;
  description: string;
  quantity: number;
  unit_price: number;
  discount: number;
  tva_rate: number;
  line_total: number;
  sort_order: number;
}

export interface InvoiceLinePayload {
  catalogue_id: number | null;
  description: string;
  quantity: number;
  unit_price: number;
  discount: number;
  tva_rate: number;
  sort_order: number;
}

export interface InvoiceSummary {
  id: number;
  number: string;
  client_id: number;
  client_name: string;
  status: InvoiceStatus;
  issue_date: string;
  due_date: string | null;
  total: number;
  amount_paid: number;
  created_at: string;
}

export interface InvoiceDetail {
  id: number;
  number: string;
  client_id: number;
  client_name: string;
  quote_id: number | null;
  quote_number: string | null;
  status: InvoiceStatus;
  issue_date: string;
  due_date: string | null;
  notes: string;
  subtotal: number;
  tva_amount: number;
  total: number;
  amount_paid: number;
  payment_method: string | null;
  payment_date: string | null;
  lines: InvoiceLine[];
  created_at: string;
  updated_at: string;
}

export interface CreateInvoicePayload {
  client_id: number;
  status?: string;
  issue_date: string;
  due_date: string | null;
  payment_method?: string;
  notes: string;
  lines: InvoiceLinePayload[];
}

export interface UpdateInvoicePayload {
  client_id: number;
  status?: string;
  issue_date: string;
  due_date: string | null;
  payment_method?: string;
  notes: string;
  lines: InvoiceLinePayload[];
}

export interface DirectSalePayload {
  client_id: number;
  issue_date: string;
  payment_method: string;
  notes: string;
  lines: InvoiceLinePayload[];
}

export interface RecordPaymentPayload {
  amount_paid: number;
  payment_method: string;
  payment_date: string;
}

// ── API calls ────────────────────────────────────

export async function getAllInvoices(
  statusFilter?: string,
): Promise<InvoiceSummary[]> {
  return invoke<InvoiceSummary[]>("get_all_invoices", {
    statusFilter: statusFilter ?? null,
  });
}

export async function getInvoiceById(id: number): Promise<InvoiceDetail> {
  return invoke<InvoiceDetail>("get_invoice_by_id", { id });
}

export async function createInvoice(
  payload: CreateInvoicePayload,
): Promise<InvoiceDetail> {
  return invoke<InvoiceDetail>("create_invoice", { payload });
}

export async function updateInvoice(
  id: number,
  payload: UpdateInvoicePayload,
): Promise<InvoiceDetail> {
  return invoke<InvoiceDetail>("update_invoice", { id, payload });
}

export async function updateInvoiceStatus(
  id: number,
  status: string,
): Promise<InvoiceDetail> {
  return invoke<InvoiceDetail>("update_invoice_status", { id, status });
}

export async function recordPayment(
  id: number,
  payload: RecordPaymentPayload,
): Promise<InvoiceDetail> {
  return invoke<InvoiceDetail>("record_payment", { id, payload });
}

export async function createDirectSale(
  payload: DirectSalePayload,
): Promise<InvoiceDetail> {
  return invoke<InvoiceDetail>("create_direct_sale", { payload });
}

export async function deleteInvoice(id: number): Promise<void> {
  return invoke<void>("delete_invoice", { id });
}

File: app/src-tauri/src/commands/invoices.rs
use tauri::State;

use crate::database;
use crate::models::invoice::*;
use crate::AppState;

#[tauri::command]
pub fn get_all_invoices(
    state: State<AppState>,
    status_filter: Option<String>,
) -> Result<Vec<InvoiceSummary>, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::get_all_invoices(&conn, status_filter.as_deref())
}

#[tauri::command]
pub fn get_invoice_by_id(state: State<AppState>, id: i64) -> Result<InvoiceDetail, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::get_invoice_detail(&conn, id)
}

#[tauri::command]
pub fn create_invoice(
    state: State<AppState>,
    payload: CreateInvoicePayload,
) -> Result<InvoiceDetail, String> {
    let mut conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::create_invoice(&mut conn, &payload)
}

#[tauri::command]
pub fn update_invoice(
    state: State<AppState>,
    id: i64,
    payload: UpdateInvoicePayload,
) -> Result<InvoiceDetail, String> {
    let mut conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::update_invoice(&mut conn, id, &payload)
}

#[tauri::command]
pub fn update_invoice_status(
    state: State<AppState>,
    id: i64,
    status: String,
) -> Result<InvoiceDetail, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::update_invoice_status(&conn, id, &status)
}

#[tauri::command]
pub fn record_payment(
    state: State<AppState>,
    id: i64,
    payload: RecordPaymentPayload,
) -> Result<InvoiceDetail, String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::record_payment(&conn, id, &payload)
}

#[tauri::command]
pub fn create_direct_sale(
    state: State<AppState>,
    payload: DirectSalePayload,
) -> Result<InvoiceDetail, String> {
    let mut conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::create_direct_sale(&mut conn, &payload)
}

#[tauri::command]
pub fn delete_invoice(state: State<AppState>, id: i64) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| format!("Lock error: {e}"))?;
    database::invoices::delete_invoice(&conn, id)
}
